src/DataTypes/Utils.cpp

```diff
@@ -0,0 +1,228 @@
+#include <DataTypes/Utils.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeTuple.h>
+
+namespace DB
+{
+
+bool canBeSafelyCasted(const DataTypePtr & from_type, const DataTypePtr & to_type)
+{
+    std::vector<std::pair<DataTypePtr, DataTypePtr>> types_to_compare;
+    types_to_compare.emplace_back(from_type, to_type);
+
+    while (!types_to_compare.empty())
+    {
+        auto [lhs_type, rhs_type] = types_to_compare.back();
+        types_to_compare.pop_back();
+
+        auto lhs_which_type = WhichDataType(lhs_type->getTypeId());
+        bool rhs_type_was_nullable = isNullableOrLowCardinalityNullable(rhs_type);
+        rhs_type = removeNullable(removeLowCardinality(rhs_type));
+
+        if (lhs_type->equals(*rhs_type))
+            continue;
+
+        auto rhs_which_type = WhichDataType(rhs_type->getTypeId());
+
+        switch (lhs_which_type.idx)
+        {
+            case TypeIndex::UInt8:
+            case TypeIndex::UInt16:
+            case TypeIndex::UInt32:
+            case TypeIndex::UInt64:
+            case TypeIndex::UInt128:
+            case TypeIndex::UInt256:
+            {
+                if (rhs_which_type.isUInt())
+                    return rhs_type->getSizeOfValueInMemory() >= lhs_type->getSizeOfValueInMemory();
+
+                if (rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::Int8:
+            case TypeIndex::Int16:
+            case TypeIndex::Int32:
+            case TypeIndex::Int64:
+            case TypeIndex::Int128:
+            case TypeIndex::Int256: {
+                if (rhs_which_type.isInt())
+                    return rhs_type->getSizeOfValueInMemory() >= lhs_type->getSizeOfValueInMemory();
+
+                if (rhs_which_type.isString())
```

Comment:
Maybe check `rhs_which_type.isString()` in the beginning and remove all this duplicating checks?

Reply:
It seems that for such complex functions it is better to make each case self contained, and avoid generalization, so we can avoid errors if we will want to add new cases.